File: lab spring25/assignment5/taskF.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

#define rept(i,n) for(int i=0;i<n;i++)
#define vi vector<int>
#define vb vector<bool>
#define br cout << '\n'

int R,C,mx;

bool valid(int r, int c){
    return 0<=r && r<R && 0<=c && c<C;
}

void bfs(string grid[],vector<vb> &vst,int r, int c){
    int dr[4] = {-1,0,1,0};
    int dc[4] = {0,1,0,-1};
    queue<int> Qr,Qc; Qr.push(r); Qc.push(c);
    vst[r][c] = true;
    int total = grid[r][c]=='D';

    while (!Qr.empty()){
        int atr = Qr.front(), atc = Qc.front(); Qr.pop(); Qc.pop();
        rept(i,4){
            int rr = atr+dr[i], cc = atc+dc[i];
            if(valid(rr,cc) && grid[rr][cc]!='#' && !vst[rr][cc]){
                total+=(grid[rr][cc]=='D');
                Qr.push(rr); Qc.push(cc);
                vst[rr][cc] = true;
            }
        }
    }
    mx = max(total,mx);
}

void solve(){
    cin >> R >> C;
    string grid[R];
    rept(i,C) cin >> grid[i];

    mx = 0;
    vector<vb> vst(R,vb(C,false));
    rept(i,R) rept(j,C) if(!vst[i][j] && grid[i][j]!='#') bfs(grid,vst,i,j);
    // rept(i,R) {
    //     rept(j,C){
    //         cout << vst[i][j];
    //     }br;
    // }
    cout << mx;br;
}

int main(){
    ios::sync_with_stdio(false);cin.tie(nullptr);cout.tie(nullptr);
    freopen("input.txt","r",stdin);
    int t{1};
    // cin >> t;
    while(t--) solve();
}

File: lab spring25/python a5/taskF.py
import sys
sys.stdin = open("input.txt","r")

mx = 0
r,c=0,0
def valid(rr,cc):
    return 0<=rr<r and 0<=cc<c

def bfs(grid,vst,R,C):
    drc = [(-1,0),(0,1),(1,0),(0,-1)]
    global mx,r,c
    Q = [(R,C)]
    vst[R][C] = 1
    total = 1
    while len(Q)!=0:
        i,j = Q.pop(0)
        for ii,jj in drc:
            rr=i+ii
            cc=j+jj
            if(valid(rr,cc) and grid[rr][cc]!="#" and vst[rr][cc]==0):
                vst[rr][cc] = 1
                Q.append((rr,cc))
                if grid[rr][cc] == "D": total+=1
    
    mx = max(total,mx)


def solve():
    global mx,r,c
    mx = 0
    r,c = list(map(int,input().split()))
    grid = list()
    for _ in range(r):
        grid.append(input())

    vst = [[0]*c for _ in range(r)]
    for i in range(r):
        for j in range(c):
            if grid[i][j] == 'D' and vst[i][j]==0: bfs(grid,vst,i,j)
    
    print(mx)

t=1
t = int(input())
while t!=0:
    t-=1
    solve()